src/main/kotlin/org/rust/ide/inspections/lints/RsPathUsageAnalysis.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections.lints
+
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.CachedValue
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.infer.ResolvedPath
+import org.rust.lang.core.types.inference
+import org.rust.openapiext.processElementsWithMacros
+
+data class PathUsageMap(
+    val pathUsages: Map<String, Set<RsElement>>,
+    val traitUsages: Set<RsTraitItem>
+)
+
+private val PATH_USAGE_KEY: Key<CachedValue<PathUsageMap>> = Key.create("PATH_USAGE_KEY")
+
+val RsItemsOwner.pathUsage: PathUsageMap
+    get() = CachedValuesManager.getCachedValue(this, PATH_USAGE_KEY) {
+        val usages = calculatePathUsages(this)
+        CachedValueProvider.Result.create(usages, PsiModificationTracker.MODIFICATION_COUNT)
+    }
+
+private fun calculatePathUsages(owner: RsItemsOwner): PathUsageMap {
+    val directUsages = hashMapOf<String, MutableSet<RsElement>>()
+    val traitUsages = hashSetOf<RsTraitItem>()
+
+    for (child in owner.children) {
+        processElementsWithMacros(child) { element ->
+            handleElement(element, directUsages, traitUsages)
+        }
+    }
+
+    return PathUsageMap(directUsages, traitUsages)
+}
+
+private fun handleElement(
+    element: PsiElement,
+    directUsages: MutableMap<String, MutableSet<RsElement>>,
+    traitUsages: MutableSet<RsTraitItem>
+): Boolean {
+    fun addItem(name: String, item: RsElement) {
+        directUsages.getOrPut(name) { hashSetOf() }.add(item)
+    }
+
+    if (!element.isEnabledByCfg) return false
+
+    return when (element) {
+        is RsModItem -> false
+        is RsPatIdent -> {
+            val name = element.patBinding.referenceName
+            val targets = element.patBinding.reference.multiResolve()
+            targets.forEach {
+                addItem(name, it)
+            }
+            true
+        }
+        is RsPath -> {
+            if (element.qualifier != null) {
+                val requiredTraits = getAssociatedItemRequiredTraits(element).orEmpty()
+                traitUsages.addAll(requiredTraits)
+            } else {
+                val useSpeck = element.parent as? RsUseSpeck
+                // `use S as T;` is also a use of S
+                if (useSpeck == null || useSpeck.alias != null) {
```

Comment:
> I changed path traversal so that use items are not special cased and decided on the following rule:
> A (root path, without a qualifier) path is a use if it
> 
> 1. isn't inside a use speck
> 2. is inside a use speck with an alias

It is a bit more complicated, e.g. there can be absolute paths (`use ::{foo1::foo2, ...}`), and also such test:

```rust
mod inner {
    pub fn func() {}
}
mod test {
    use crate::inner;
    use inner::{func};
    fn main() {
        func();
    }
}
```

I would do something like this:
```suggestion
                val useSpeck = element.parentOfType<RsUseSpeck>()
                if (useSpeck == null || useSpeck.isTopLevel) {


/**
 * We should collect paths only from relative use specks,
 * that is top-level use specks without `::`
 * E.g. we shouldn't collect such paths: `use ::{foo, bar}`
 */
private val RsUseSpeck.isRelative: Boolean
    get() = (path != null || coloncolon == null)
        && parentOfType<RsUseSpeck>()?.isRelative != false

```

Reply:
Oh, I see that my way worked "by accident". Thanks for the test.